File: frontend/src/utils/validation.ts
import type { PasswordStrength } from "../types";

export const validateUsername = (
  username: string
): { isValid: boolean; error?: string } => {
  const trimmed = username.trim();

  if (trimmed.length < 3) {
    return {
      isValid: false,
      error: "Username must be at least 3 characters long",
    };
  }

  if (trimmed.length > 30) {
    return { isValid: false, error: "Username must not exceed 30 characters" };
  }

  const usernameRegex = /^[a-zA-Z0-9_-]+$/;
  if (!usernameRegex.test(trimmed)) {
    return {
      isValid: false,
      error:
        "Username can only contain letters, numbers, underscores, and hyphens",
    };
  }

  return { isValid: true };
};

export const calculatePasswordStrength = (
  password: string
): PasswordStrength => {
  if (!password) {
    return {
      score: 0,
      label: "",
      color: "#6b7280",
      requirements: {
        length: false,
        uppercase: false,
        lowercase: false,
        number: false,
        special: false,
      },
    };
  }

  const requirements = {
    length: password.length >= 8,
    uppercase: /[A-Z]/.test(password),
    lowercase: /[a-z]/.test(password),
    number: /[0-9]/.test(password),
    special: /[^A-Za-z0-9]/.test(password),
  };

  let score = 0;
  if (requirements.length) score++;
  if (requirements.uppercase) score++;
  if (requirements.lowercase) score++;
  if (requirements.number) score++;
  if (requirements.special) score++;

  let label: string;
  let color: string;

  if (score <= 2) {
    label = "Weak";
    color = "#ef4444";
  } else if (score === 3) {
    label = "Fair";
    color = "#f97316";
  } else if (score === 4) {
    label = "Good";
    color = "#3b82f6";
  } else {
    label = "Strong";
    color = "#10b981";
  }

  return {
    score,
    label,
    color,
    requirements,
  };
};

export const validatePassword = (
  password: string
): { isValid: boolean; error?: string } => {
  if (password.length < 8) {
    return {
      isValid: false,
      error: "Password must be at least 8 characters long",
    };
  }

  if (password.length > 100) {
    return { isValid: false, error: "Password must not exceed 100 characters" };
  }

  const hasUpper = /[A-Z]/.test(password);
  const hasLower = /[a-z]/.test(password);
  const hasNumber = /[0-9]/.test(password);

  if (!hasUpper || !hasLower || !hasNumber) {
    return {
      isValid: false,
      error:
        "Password must contain at least one uppercase letter, one lowercase letter, and one number",
    };
  }

  return { isValid: true };
};

export const sanitizeMessage = (message: string): string => {
  return message
    .trim()
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#x27;")
    .replace(/\//g, "&#x2F;");
};

export const validateRoomName = (
  roomName: string
): { isValid: boolean; error?: string } => {
  const trimmed = roomName.trim();

  if (!trimmed) {
    return { isValid: false, error: "Room name cannot be empty" };
  }

  if (trimmed.length > 50) {
    return { isValid: false, error: "Room name must not exceed 50 characters" };
  }

  return { isValid: true };
};

File: src/middleware/validation.go
package middleware

import (
	"html"
	"regexp"
	"strings"
	"unicode"

	"github.com/gofiber/fiber/v2"
)

// Validation rules
const (
	MinUsernameLength = 3
	MaxUsernameLength = 30
	MinPasswordLength = 8
	MaxPasswordLength = 100
	MaxMessageLength  = 2000
	MaxRoomNameLength = 50
)

// Username validation regex (alphanumeric, underscore, hyphen)
var usernameRegex = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

// ValidateUsername validates username format and length
func ValidateUsername(username string) (string, error) {
	// Trim whitespace
	username = strings.TrimSpace(username)

	// Check length
	if len(username) < MinUsernameLength {
		return "", fiber.NewError(fiber.StatusBadRequest, 
			"Username must be at least 3 characters long")
	}
	if len(username) > MaxUsernameLength {
		return "", fiber.NewError(fiber.StatusBadRequest, 
			"Username must not exceed 30 characters")
	}

	// Check format (alphanumeric, underscore, hyphen only)
	if !usernameRegex.MatchString(username) {
		return "", fiber.NewError(fiber.StatusBadRequest, 
			"Username can only contain letters, numbers, underscores, and hyphens")
	}

	// Sanitize (escape HTML)
	username = html.EscapeString(username)

	return username, nil
}

// ValidatePassword validates password strength
func ValidatePassword(password string) error {
	// Check length
	if len(password) < MinPasswordLength {
		return fiber.NewError(fiber.StatusBadRequest, 
			"Password must be at least 8 characters long")
	}
	if len(password) > MaxPasswordLength {
		return fiber.NewError(fiber.StatusBadRequest, 
			"Password must not exceed 100 characters")
	}

	// Check for at least one uppercase letter
	hasUpper := false
	// Check for at least one lowercase letter
	hasLower := false
	// Check for at least one digit
	hasDigit := false

	for _, char := range password {
		switch {
		case unicode.IsUpper(char):
			hasUpper = true
		case unicode.IsLower(char):
			hasLower = true
		case unicode.IsDigit(char):
			hasDigit = true
		}
	}

	if !hasUpper {
		return fiber.NewError(fiber.StatusBadRequest, 
			"Password must contain at least one uppercase letter")
	}
	if !hasLower {
		return fiber.NewError(fiber.StatusBadRequest, 
			"Password must contain at least one lowercase letter")
	}
	if !hasDigit {
		return fiber.NewError(fiber.StatusBadRequest, 
			"Password must contain at least one number")
	}

	return nil
}

// SanitizeMessage sanitizes message content
func SanitizeMessage(message string) (string, error) {
	// Trim whitespace
	message = strings.TrimSpace(message)

	// Check if empty
	if message == "" {
		return "", fiber.NewError(fiber.StatusBadRequest, 
			"Message cannot be empty")
	}

	// Check length
	if len(message) > MaxMessageLength {
		return "", fiber.NewError(fiber.StatusBadRequest, 
			"Message must not exceed 2000 characters")
	}

	// Escape HTML to prevent XSS
	message = html.EscapeString(message)

	return message, nil
}

// ValidateRoomName validates room name
func ValidateRoomName(roomName string) (string, error) {
	// Trim whitespace
	roomName = strings.TrimSpace(roomName)

	// Check if empty
	if roomName == "" {
		return "", fiber.NewError(fiber.StatusBadRequest, 
			"Room name cannot be empty")
	}

	// Check length
	if len(roomName) > MaxRoomNameLength {
		return "", fiber.NewError(fiber.StatusBadRequest, 
			"Room name must not exceed 50 characters")
	}

	// Sanitize (escape HTML)
	roomName = html.EscapeString(roomName)

	return roomName, nil
}

// SanitizeInput is a general purpose input sanitizer
func SanitizeInput(input string, maxLength int) string {
	// Trim whitespace
	input = strings.TrimSpace(input)

	// Truncate if too long
	if len(input) > maxLength {
		input = input[:maxLength]
	}

	// Escape HTML
	input = html.EscapeString(input)

	return input
}
